In [2]:
#hiddencell
from pbl_tools import *

# 스테이지 6

# 1.NumPy 라이브러리(Library) 불러오기
[문제 1]    
numpy 라이브러리를 불러오기 위해서 아래 빈칸을 채워 주세요.    
numpy는 자주 사용하는 라이브러리이므로 축약어인 np로 불러와 주세요. 

In [4]:
#checkcode
ensure_vals(globals(), 'np')
@check_safety
def check(user_answer_pandas=np):
    if hasattr(user_answer_pandas, 'array'):
        return True
    else:
        return False
check()

True

### Inst.
이 코드는 NumPy 라이브러리(Library)를 축약어인 np로 불러오는 기법입니다.

### Hint.
np는 numpy의 약칭입니다.   
import 와 as를 이용해서 numpy 라이브러리(Library)를 불러와주세요.

### Solution.
```python
import numpy as np
```

# 2.누적 합(cumsum) 구하기

In [5]:
two_dimension = np.array([[5,10],
                          [15,20]])

np.cumsum(two_dimension)

array([ 5, 15, 30, 50])

In [7]:
#checkcode
#empty

### Inst.
   
해당 코드는 NumPy의 `cumsum()` 함수를 사용하여 **누적 합(cumulative sum)** 을 구하는 코드입니다.

`cumsum()` 함수는 인자로 받은 **배열의 요소들을 누적하여 합** 을 구한 배열을 반환합니다.   
이 함수는 `axis` 옵션을 사용하여 **각 차원별로 누적 합** 을 계산할 수 있습니다.   
axis의 기본값은 `None` 입니다. `None` 으로 지정할 경우 **전체 배열을 대상으로 연산이 수행**됩니다.  

해당 코드에서는 two_dimension 배열에 대해서 누적 합을 계산하고 있습니다.  
`cumsum()` 함수는 two_dimension 배열의 각 요소를 **왼쪽 상단부터 오른쪽 하단까지** 순서대로 더한 값을 배열로 반환합니다. 

<img width="300" src="https://drive.google.com/uc?id=1Af-mjU3-L65lczSV4vYX65Ar-T8kZL2Q">

즉, 반환된 배열의 첫 번째 요소는 two_dimension 배열의 **첫 번째** 요소,  
두 번째 요소는 two_dimension 배열의 **첫 번째 요소와 두 번째** 요소를 더한 값,  
세 번째 요소는 two_dimension 배열의 **첫 번째, 두 번째, 세 번째** 요소를 더한 값,  
네 번째 요소는 two_dimension 배열의 **모든 요소**를 더한 값이 됩니다.  

`cumsum()` 함수는 NumPy에서 누적 합을 구할 때 유용하게 사용됩니다.   
예를 들어, 주식 시장에서 일정 기간동안 주식 가격이 어떻게 변화했는지 분석할 때,   
누적 합을 구하여 총 수익률을 계산하는 등의 용도로 활용될 수 있습니다.

---

#### 결과 해석
two_dimension 배열의 각 요소를 왼쪽 상단부터 오른쪽 하단까지 더한 결과인 **[ 5, 15, 30, 50]** 를 반환합니다. 

### Hint.
empty

### Solution.
empty

# 3.NumPy 배열의 축(axis)을 기준으로 누적 합(cumsum) 구하기

In [13]:
two_dimension = np.array([[5,10],
                          [15,20]])

np.cumsum(two_dimension, axis = 1)

array([[ 5, 15],
       [15, 35]])

In [ ]:
#checkcode
#empty

### Inst.

해당 코드는 NumPy의 `cumsum()` 함수를 사용하여 누적 합을 구하는 코드입니다.

`cumsum()` 함수는 주어진 배열의 요소들의 누적 합을 반환합니다.   
두 번째 인자 **axis를 지정하여 연산하려는 축을 선택** 할 수 있습니다.  
`axis = 1`일 경우, 수평 방향의 누적 합을 계산하며,   
`axis = 0`일 경우, 수직 방향의 누적 합을 계산합니다.  


위 코드에서는 `axis=1` 을 지정하여 **수평 방향으로 누적 합** 을 계산합니다.   
따라서, 결과적으로는 다음과 같은 배열이 반환됩니다.

<img width="282" src="https://drive.google.com/uc?id=1L65E8ajg9UI2o6l3P2uhKoIveCdnX43W">

---

#### 결과 해석
첫 번째 행은 [5, 10]의 누적 합인 [5, 15]가 되고, 두 번째 행은 [15, 20]의 누적 합인 [15, 35]가 됩니다.

### Hint.
empty

### Solution.
empty

# 4.NumPy 배열의 축(axis)을 기준으로 누적 합(cumsum) 구하기 실습
[문제 2]  
two_dimension 배열에서 수직 방향으로 누적 합을 계산해 보세요.

In [10]:
two_dimension = np.array([[5,10],
                          [15,20]])

np.___(___, axis = ___)

array([[ 5, 10],
       [20, 30]])

In [11]:
#checkcode
ensure_vals(globals(),'two_dimension')
@check_safety
def check(user_answer=_):
    check1 = (user_answer == np.cumsum(two_dimension, axis = 0)).all()
    if check1:
        return True
    else:
        return False
check()

True

### Inst.
`cumsum()` 함수는 인자로 받은 **배열의 요소들을 누적하여 합** 을 구한 배열을 반환합니다.   
이 함수는 두 번째 인자 **axis를 지정하여 연산하려는 축을 선택** 할 수 있습니다.

---

#### 결과 해석
two_dimension 배열의 수직 방향 누적합 **[[5,10], [20,30]]** 가 반환됩니다.  

### Hint.
`cumsum()`, `axis = 0` 함수를 사용하면 수직 방향의 누적 합을 구할 수 있습니다.

### Solution.
```python
np.cumsum(two_dimension, axis = 0)
`````

# 5.e의 지수함수(exp) 적용

In [16]:
np.exp(2)

7.38905609893065

In [ ]:
#checkcode
#empty

### Inst.

해당 코드의 `exp()` 함수는 NumPy에서 **지수 함수(exponential function)** 를 계산하기 위해 사용됩니다.   
`exp()` 함수는 NumPy 배열에서 **요소별(element-wise) 지수 계산** 을 수행합니다.

위 코드에서 `np.exp(2)`는 NumPy 라이브러리의 exp 함수를 이용하여 2의 지수 함수 값을 계산하는 코드입니다.   
exp 함수는 입력된 값의 지수값을 계산하는 함수로, 자연 상수(e)의 지수 함수 값을 계산하는 것과 같습니다.

---

#### 결과 해석

np.exp(2)는 e의 2승, 즉 7.3890560989306495 값을 반환합니다.

### Hint.
empty

### Solution.
empty

# 6.e의 지수함수(exp) 적용 실습
[문제 3]   
[1,2,3,4] 리스트 배열에 대해 각 요소에 대한 지수 값을 계산해보세요.  
아래 빈칸을 채워보세요.

In [16]:
np.___(___)   

array([ 2.71828183,  7.3890561 , 20.08553692, 54.59815003])

In [17]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_):
    check1 = (user_answer == np.exp([1,2,3,4])).all()
    if check1:
        return True
    else:
        return False
check()

True

### Inst.

해당 코드의 `exp()` 함수는 NumPy 배열에서 **요소별(element-wise) 지수 계산** 을 수행합니다.  
파라미터에는 간단한 숫자뿐만 아니라 **list 형태로도 삽입** 이 될 수 있습니다.  

---

#### 결과 해석
[1,2,3,4]리스트에 지수함수를 적용한 결과 **[ 2.71828183,  7.3890561 , 20.08553692, 54.59815003]** 가 반환됩니다.  

### Hint.
`exp()` 함수를 사용하면 각 요소에 대한 지수 값을 구할 수 있습니다.

### Solution.
```python
np.exp([1,2,3,4])
```

# 7.NumPy 배열의 제곱근(sqrt) 계산하기

In [18]:
np.sqrt(4)

2.0

In [ ]:
#checkcode
#empty

### Inst.
`sqrt()` 함수는 **제곱근** 을 계산해 줍니다.   
exp 함수와 마찬가지로 파라미터에는 **간단한 숫자뿐만 아니라 list 형태로도 삽입**이 될 수 있습니다.

해당 코드는 NumPy의 `sqrt()` 함수를 사용하여 **인자로 전달된 값을 제곱근으로 반환** 합니다.

이 코드는 숫자 4의 제곱근을 구하는 코드입니다.   
np.sqrt(4)는 2를 반환합니다.

`sqrt()` 함수는 NumPy의 벡터화 연산을 지원하기 때문에,   
배열 또는 리스트와 같은 여러 값을 인자로 전달하여 한 번에 제곱근을 구할 수도 있습니다.   
예를 들어, np.sqrt([1, 4, 9, 16])는 배열 [1, 2, 3, 4]를 반환합니다.  

---

#### 결과 해석

코드의 실행결과 4의 제곱근 2.0이 반환됩니다.  

### Hint.
empty

### Solution.
empty

# 8.NumPy 배열의 제곱근(sqrt) 계산하기 실습

[문제 4]  
4,9,16을 리스트 형태로 삽입하여 제곱근을 출력해보세요.

In [19]:
np.___(___)

array([2., 3., 4.])

In [20]:
#checkcode
ensure_vals(globals())
@check_safety
def check(user_answer=_):
    check1 = (user_answer == np.sqrt([4,9,16])).all()
    if check1:
        return True
    else:
        return False
check()

True

### Inst.
`sqrt()` 함수는 제곱근을 계산할 수 있습니다.   
파라미터에는 간단한 숫자뿐만 아니라 list 형태로도 삽입이 될 수 있습니다.  

---

#### 결과 해석
\[4,9,16\] 리스트에 제곱근을 구하는 넘파이 함수를 사용하면 **[2., 3., 4.]** 가 반환됩니다.  

### Hint.
`sqrt()` 함수를 사용하면 각 요소에 대한 제곱근 값을 구할 수 있습니다.  
list의 형태는 `[4, 9, 16]` 입니다.

### Solution.
```python
np.sqrt([4,9,16])
```